File: backend/library/collection_tree.py
"""The collection tree: nested collections (folders) as a flat list plus ``parent_id``.

docs/plans/library-finder.md §2. Pure and model-free: every helper takes anything
with ``id``, ``name`` and ``parent_id`` (a :class:`~collection_store.Collection`),
so ``collection_store`` can import this module and never the other way round.

Rules this module holds:

* **Depth counts levels from the root**: a top-level collection is depth 1, and
  a move is judged by the depth its *deepest* descendant would land at
  (``depth_of(parent) + subtree_height(moving)``), so a whole subtree fits or
  none of it moves.
* **The file must be a forest**: a dangling ``parent_id`` or a cycle makes the
  file unreadable (:func:`tree_problem`), never a silently flattened tree.
* Every walk carries a visited set, so even a list that was never validated
  cannot loop forever.
"""

from __future__ import annotations

from typing import Mapping, Optional, Protocol, Sequence, TypeVar

from backend.library.errors import CollectionCycle, CollectionTooDeep, UnknownParent

#: How many levels collections may nest; a top-level collection is level 1.
MAX_COLLECTION_DEPTH = 8


class TreeNode(Protocol):
    @property
    def id(self) -> str: ...
    @property
    def name(self) -> str: ...
    @property
    def parent_id(self) -> Optional[str]: ...


Node = TypeVar("Node", bound=TreeNode)


def _by_id(nodes: Sequence[Node]) -> dict[str, Node]:
    return {node.id: node for node in nodes}


def ancestry(nodes: Sequence[Node], node_id: str) -> tuple[Node, ...]:
    """The chain from the top-level ancestor down to ``node_id``, itself last.

    ``()`` for an unknown id. A dangling parent ends the chain where it breaks;
    a cycle ends it before the repeat (a validated file has neither).
    """
    index = _by_id(nodes)
    chain: list[Node] = []
    seen: set[str] = set()
    current = index.get(node_id)
    while current is not None and current.id not in seen:
        seen.add(current.id)
        chain.append(current)
        current = index.get(current.parent_id) if current.parent_id is not None else None
    return tuple(reversed(chain))


def children_of(nodes: Sequence[Node], node_id: str) -> tuple[Node, ...]:
    """The direct subfolders of ``node_id``, in list order."""
    return tuple(node for node in nodes if node.parent_id == node_id)


def descendant_ids(nodes: Sequence[TreeNode], node_id: str) -> frozenset[str]:
    """Every id below ``node_id`` at any depth, never ``node_id`` itself."""
    found: set[str] = set()
    frontier = [node_id]
    while frontier:
        parent = frontier.pop()
        for child in children_of(nodes, parent):
            if child.id not in found and child.id != node_id:
                found.add(child.id)
                frontier.append(child.id)
    return frozenset(found)


def depth_of(nodes: Sequence[TreeNode], node_id: str) -> int:
    """1 for a top-level collection, 2 inside it…; 0 for an unknown id."""
    return len(ancestry(nodes, node_id))


def subtree_height(nodes: Sequence[TreeNode], node_id: str) -> int:
    """Levels in ``node_id``'s subtree, itself included: 1 for a leaf."""
    base = depth_of(nodes, node_id)
    deepest = max((depth_of(nodes, d) for d in descendant_ids(nodes, node_id)), default=base)
    return deepest - base + 1


def collection_path(nodes: Sequence[TreeNode], node_id: str) -> list[str]:
    """Names from the top-level ancestor down to ``node_id``; ``[]`` when unknown."""
    return [node.name for node in ancestry(nodes, node_id)]


def total_members(
    nodes: Sequence[TreeNode], member_counts: Mapping[str, int], node_id: str
) -> int:
    """Videos in ``node_id`` plus every descendant (direct membership summed)."""
    ids = {node_id, *descendant_ids(nodes, node_id)}
    return sum(member_counts.get(each, 0) for each in ids)


def check_placement(
    nodes: Sequence[TreeNode], moving_id: Optional[str], parent_id: Optional[str]
) -> None:
    """Raise when ``parent_id`` cannot hold ``moving_id`` (``None`` = a new collection).

    Checked in this order, so each refusal names the first thing wrong: an
    unknown parent, a cycle, then the depth. ``parent_id=None`` (the root) is
    always a legal place.
    """
    if parent_id is None:
        return
    if parent_id not in _by_id(nodes):
        raise UnknownParent(parent_id)
    if moving_id is not None and (
        parent_id == moving_id or parent_id in descendant_ids(nodes, moving_id)
    ):
        raise CollectionCycle(moving_id, parent_id)
    height = subtree_height(nodes, moving_id) if moving_id is not None else 1
    if depth_of(nodes, parent_id) + height > MAX_COLLECTION_DEPTH:
        raise CollectionTooDeep(parent_id, MAX_COLLECTION_DEPTH)


def tree_problem(nodes: Sequence[TreeNode]) -> Optional[str]:
    """Why ``nodes`` is not a forest (a dangling parent or a cycle), or None."""
    index = _by_id(nodes)
    for node in nodes:
        if node.parent_id is not None and node.parent_id not in index:
            return (
                f"collection {node.id!r} names a parent_id {node.parent_id!r} "
                "that does not exist"
            )
    for node in nodes:
        seen = {node.id}
        current = node.parent_id
        while current is not None:
            if current in seen:
                return f"the parent_ids above collection {node.id!r} loop back on themselves"
            seen.add(current)
            current = index[current].parent_id
    return None

File: src/renderer/src/lib/collectionTree.ts
/**
 * The collection tree (docs/plans/library-finder.md §2, §4.1): nested
 * collections — "folders" in the UI — arrive as a **flat** list where each row
 * names its `parent_id` (`null` = top level), and the renderer builds the tree.
 *
 * Every helper is generic over `{ id, parent_id }`, so Settings → Folders, the
 * library sidebar and drag-and-drop targets share one implementation. The
 * backend is the authority on where a folder may go; `canMoveInto` is only the
 * renderer's copy of its rules, used to leave a place out of a picker or to
 * grey out a drop target.
 *
 * Defensive by design: a row whose parent is unknown is shown at the top level,
 * and every walk carries a visited set, so a malformed list can never loop or
 * make a folder vanish.
 *
 * Pure module: no React, no `window`, no I/O.
 */

/** Mirrors the backend's `MAX_COLLECTION_DEPTH`: a top-level folder is depth 1. */
export const MAX_COLLECTION_DEPTH = 8

/** Between the names of a folder's path: `Events › UCK 2026 › Day 1`. */
export const PATH_SEPARATOR = ' › '

export interface TreeItem {
  id: string
  parent_id: string | null
}

export interface TreeNode<T extends TreeItem> {
  item: T
  /** 1 for a top-level folder. */
  depth: number
  children: TreeNode<T>[]
}

/** A tree node flattened into display order. */
export interface TreeRow<T extends TreeItem> {
  item: T
  depth: number
}

function indexById<T extends TreeItem>(items: readonly T[]): Map<string, T> {
  return new Map(items.map((item) => [item.id, item]))
}

/**
 * The forest, top-level folders first. Siblings keep the input order unless
 * `compare` sorts them. A row whose parent is unknown — or that is only
 * reachable through a cycle — is placed at the top level rather than dropped.
 */
export function buildTree<T extends TreeItem>(
  items: readonly T[],
  compare?: (a: T, b: T) => number
): TreeNode<T>[] {
  const index = indexById(items)
  const placed = new Set<string>()
  const sorted = (rows: T[]): T[] => (compare ? [...rows].sort(compare) : rows)

  function node(item: T, depth: number): TreeNode<T> {
    placed.add(item.id)
    const kids = items.filter((c) => c.parent_id === item.id && !placed.has(c.id))
    kids.forEach((c) => placed.add(c.id))
    return { item, depth, children: sorted(kids).map((c) => node(c, depth + 1)) }
  }

  const isRoot = (item: T) => item.parent_id === null || !index.has(item.parent_id)
  const roots = sorted(items.filter(isRoot)).map((item) => node(item, 1))
  const stranded = items.filter((item) => !placed.has(item.id))
  // A cycle has no root; show its first member at the top so nothing disappears.
  return stranded.length === 0 ? roots : [...roots, ...buildTree(breakCycles(stranded), compare)]
}

/** The stranded rows with their first member lifted to the top level. */
function breakCycles<T extends TreeItem>(stranded: readonly T[]): T[] {
  const [first, ...rest] = stranded
  return [{ ...first, parent_id: null }, ...rest]
}

/** Pre-order: each folder, then its subfolders. */
export function flattenTree<T extends TreeItem>(tree: readonly TreeNode<T>[]): TreeRow<T>[] {
  return tree.flatMap((node) => [
    { item: node.item, depth: node.depth },
    ...flattenTree(node.children),
  ])
}

/** The folders above `id`, top level first; `[]` for a top-level or unknown id. */
export function ancestorsOf<T extends TreeItem>(items: readonly T[], id: string): T[] {
  const index = indexById(items)
  const chain: T[] = []
  const seen = new Set<string>([id])
  let parentId = index.get(id)?.parent_id ?? null
  while (parentId !== null && !seen.has(parentId)) {
    const parent = index.get(parentId)
    if (!parent) break
    seen.add(parentId)
    chain.unshift(parent)
    parentId = parent.parent_id
  }
  return chain
}

/** Every id below `id` at any depth, never `id` itself. */
export function descendantIds(items: readonly TreeItem[], id: string): Set<string> {
  const found = new Set<string>()
  const frontier = [id]
  while (frontier.length > 0) {
    const parent = frontier.pop()
    for (const item of items) {
      if (item.parent_id === parent && item.id !== id && !found.has(item.id)) {
        found.add(item.id)
        frontier.push(item.id)
      }
    }
  }
  return found
}

/** 1 for a top-level folder, 2 inside it…; 0 for an unknown id. */
export function depthOf(items: readonly TreeItem[], id: string): number {
  return items.some((item) => item.id === id) ? ancestorsOf(items, id).length + 1 : 0
}

/** Levels in `id`'s subtree, itself included: 1 for a folder with no subfolders. */
export function subtreeHeight(items: readonly TreeItem[], id: string): number {
  const base = depthOf(items, id)
  const depths = [...descendantIds(items, id)].map((d) => depthOf(items, d))
  return Math.max(base, ...depths) - base + 1
}

/**
 * Whether the folder `movingId` may sit inside `targetId` (`null` = the top
 * level): the target exists, is neither the folder nor one of its subfolders,
 * and the folder's whole subtree still fits under `MAX_COLLECTION_DEPTH`.
 * Its current parent is allowed (a no-op move).
 */
export function canMoveInto(
  items: readonly TreeItem[],
  movingId: string,
  targetId: string | null
): boolean {
  if (!items.some((item) => item.id === movingId)) return false
  if (targetId === null) return true
  if (!items.some((item) => item.id === targetId)) return false
  if (targetId === movingId || descendantIds(items, movingId).has(targetId)) return false
  return depthOf(items, targetId) + subtreeHeight(items, movingId) <= MAX_COLLECTION_DEPTH
}

/**
 * The folders `movingId` may move into, in tree order with their depth — what a
 * "Location" picker or a "Move to…" menu lists. The top level is not a row:
 * `canMoveInto(items, movingId, null)` is always true for a known folder.
 */
export function moveTargets<T extends TreeItem>(
  items: readonly T[],
  movingId: string,
  compare?: (a: T, b: T) => number
): TreeRow<T>[] {
  return flattenTree(buildTree(items, compare)).filter((row) =>
    canMoveInto(items, movingId, row.item.id)
  )
}

/** Whether a new folder may be created inside `targetId` (`null` = top level). */
export function canCreateInside(items: readonly TreeItem[], targetId: string | null): boolean {
  if (targetId === null) return true
  const depth = depthOf(items, targetId)
  return depth > 0 && depth < MAX_COLLECTION_DEPTH
}

/** `Events › UCK 2026 › Day 1` — the folder's names from the top level down. */
export function pathLabel<T extends TreeItem & { name: string }>(
  items: readonly T[],
  id: string
): string {
  const self = items.find((item) => item.id === id)
  if (!self) return id
  return [...ancestorsOf(items, id), self].map((item) => item.name).join(PATH_SEPARATOR)
}
